### PR TITLE
gui: Implement GUI version of consolidateunspent (coin control part)

### LINKED ISSUE #1984: Create a GUI for the consolidateunspent Command
Currently the only way to consolidate UTXOs with the GUI is manually with the advance coin control or you have to open the console and run the `consolidateunspent` command. Since a fair number of people run into issues with having too many small UTXOs, a simple GUI for this would be helpful

### PR DESCRIPTION
**Pull request by jamescowens**
This PR extends coin control to
1. Change the (un)select all button to a true select all/select none filp-flop.
2. Implement an input filter function to allow the selection of inputs either less than or equal to or greater than or equal to a provided input value, with an automatically imposed limit of 200 inputs (to prevent transaction failures).
3. Implementation of a "Consolidate" button which (re)filters the inputs and then presents a pick list to select the destination address for the consolidation, and then automatically fills in the send transaction for convenience.

What is left is the implementation of a wizard for the consolidate unspent which overlays on top of this machinery. I would like to do that in a separate PR.

Closes #1984.

### CODE REVIEW

**Comment by jamescowens** [comment]
Here are some screenshots of a testnet node using this PR's features...

Start with the send tab (coin control turned on).
image

Push the "Inputs..." button.
image

Select the desired inputs to be subject to consolidation. Note that you can just select one of the first level nodes of the tree, which corresponds to the inputs on a particular address. (Cf. the single address mode of the rpc command.) You can also use the "(un)select all" button to select all of the inputs - this is similar to the sweep from all addresses boolean in the rpc command.
image

Notice that there are 1497 inputs selected. Next fill in the less than x GRC field to direct the dialog to select a subset of the UTXO's already selected that are equal to or less than the provided value (in this case 1100 GRC).
image

Push the "Consolidate Outputs" button.
image

The modal consolidate outputs dialog comes up and it provides a table for you to select the destination address for the consolidation. If over 200 inputs met the selection criteria, they will be ordered in increasing value and only the first 200 selected, and then a warning will appear letting you know that the selection scope has been limited to 200 to prevent transaction failure due to too many inputs.

After you select the address and press "ok". You return to the coin control dialog.
image

You notice that the number of outputs has been limited to 200, and expanding one of the branches of the tree, you can see that the actual selection value stopped at about 900 instead of 1100 because the number of inputs limit was reached.

Another branch... same thing...
image

Finally, after pressing "ok" on the coin control dialog box, you are returned to the send tab. Notice that the destination address, label, and amount are already pre-filled in along with the same address in the custom change address (in case there is a slight amount of change due to the uncertainty in the fee calculation).
image

The consolidation transaction is ready to send by pressing the "send" button.

**Comment by jamescowens** [comment]
Note that this integrated with coin control approach takes longer to do than the rpc command, but it offers more flexibility, which is appropriate for the GUI. It also (not shown above) allows things that are not supported by the current rpc consolidateunspent function, such as fine tuning the inputs selected on the coin control after the consolidate selection filter is applied but before sending, or selecting the inputs for consolidation from one address, but using a destination address that is different. (That is not currently supported by the rpc command.)

I know this needs some polishing, but it is good enough to review I think.

**Comment by presciencia** [comment]
It looks like better for the wizard. I will need to see this instruction each time I use this 😄

**Comment by jamescowens** [comment]
It is not as hard as you think once you try it. And for you to use consolidateunspent, you have to know what a UTXO is, which means you should also be familiar with coin control and the inputs button. We can turn the Consolidate Outputs into a bonafide wizard if we want. Maybe we should also just say Consolidate, because people get confused by the outputs and inputs thing too.

**Comment by cyrossignol** [comment]
I think it could be really effective in a wizard-like flow. It feels a bit unnatural to start this process from within the coin picker because the UI controls will have a dual meaning. Though it doesn't seem hard to figure out from my perspective, it's not immediately obvious from the dialog how to use the feature and that the outputs need to be selected first. 

The folks who may benefit most from this tool are probably not experts with UTXOs and coin control. These are concepts that some struggle to grasp, so a targeted UI version of the consolidator could have a much greater impact on the people without the experience&mdash;those who know how to do coin control can already create these transactions using the existing GUI and RPC tools without trouble (I usually just check-off an address in the selector and address book, copy/paste amount, and press send). The widgets are already done here&mdash;it doesn't seem hard to reorganize them into a wizard.

> Maybe we should also just say Consolidate, because people get confused by the outputs and inputs thing too.

I like that better. Maybe "Filter" instead?

I wonder... could we reuse the amount filter for regular coin selection as well? It seems like a nice feature that we'd get for free along with this.

**Comment by jamescowens** [comment]
The wizard could simply use this machinery (underlying), much like I am actually driving the underlying senddialog. And filter is a better word. The only things specific about the consolidation filter are the limits on inputs and the selection of a destination address. Seems like a good idea to have it as a generalized feature for regular coin selection.

**Comment by jamescowens** [comment]
I still think we have no business presenting the wizard for UTXO consolidation unless advanced coin control is enabled. For you to ask the question as to whether you need to consolidate means that you have to understand what an input/output is. If you don't then the concept of consolidate UTXO's doesn't mean anything - unless we want the wizard to truly be a tutorial.

BTW this is, as I feared when starting it, turning into a larger project than I hoped, because we will descend into arguments on how it should look and function.

I do NOT like the idea of simply "wizardizing" the parameters in the RPC function, because people that don't know those parameters are going to be clueless, even in a wizard, on what to put in there.

**Comment by RoboticMind** [comment]
Partly due to the existing clutter in the coin control interface, I think some of the layout may be a bit hard to follow. I had a harder time seeing how things worked and I was aware of what UTXOs are and all. 

Maybe something like one of those below would be better? Ignore the weird spacing/weird text since I made this quickly

image

or like 

image

or something else. 

Have a couple other suggestions like greying out the consolidate button if nothing is selected and making the (un)select all just switch between `select all` and `unselect all` since it took me a minute to realize that it meant that it would either select or unselect depending on whether you had selecting things or not


No idea on the amount of work required for this, so let me know if this is too much

**Comment by jamescowens** [comment]
I have separated out the filter into a separate function from the GUI button to facilitate more generic use.

**Comment by vincehusky** [comment]
> BTW this is, as I feared when starting it, turning into a larger project than I hoped, because we will descend into arguments on how it should look and function.

Look and function...if this is not the whole essence of UI design, what else is there?

Apologies. I could not resist being a smart---; no disrespect intended. In fact I am thankful for your efforts. As I am a UI designer, I have started watching the Gridcoin UI updates. Liking the progress!

At risk of being a nuisance for you all, just want to add to the discusion since I am confused by the presented functionality. I do not frequently use the coin control so perhaps I just don't "get it." I am at a point in my knowledge where I get utxos and consolidating them for staking, but it's not carrying over to what it means above. Some others said "not obvious" so the question could be what can make it obvious. A good UI can make complex ideas simple and clear. I think I fall into the group of users who would benefit from some coddling.

I am no good for code but I wish I could help contribute to this great project. I hope this is some food for thought. If there are UI related tasks that I can assist with please give a shout.

**Comment by jamescowens** [comment]
Ok. @cyrossignol and folks. Take a look at this in a partially completed state. I have not done the consolidate button over as a wizard yet, so ignore that one, but I changed the behavior and wording on the select all/select none button to be a true toggle. I have also rearranged the filter to be more logical. It is now generalized, so you can select the outputs either less than or equal to or greater than or equal to the filled in GRC value.
image

**Comment by cyrossignol** [comment]
@jamescowens Hehe, that's one big button 🙂

I think the filter looks great. How does the "greater than or equal to" button work? Does it cycle through the options with each click? We may want to use the `>=`, etc., symbols instead of text to avoid the work for translators.

**Comment by jamescowens** [comment]
Haha. Ignore the appearance of the buttons. I will try and put spacers in there to make it more reasonable. It is a flip flop. It flips from less than (the default) to greater than and back as you repeatedly press it.

**Comment by jamescowens** [comment]
I was thinking about using those symbols too, but will everyone know what <= or >= mean. (Using the same thought that this area has been too wonky....)

**Comment by jamescowens** [comment]
Ok. This looks more reasonable...
image

**Comment by cyrossignol** [comment]
Yep. I like it.

**Comment by cyrossignol** [comment]
> I was thinking about using those symbols too, but will everyone know what <= or >= mean.

Perhaps some won't. I think the symbols are common enough for most people to grasp, though. At least for the time being, these should cover more ground until translations catch up.

**Comment by jamescowens** [comment]
Ok. I think this is now pretty straightforward and logical. This wraps up the coin control part. I will do a wizard next.

Here is the basic flow. You should actually compile this and use it to try it out for yourself. There are good explanatory tooltips for the buttons now.

Start with coin control...
image

Select desired coins... (in this case I pressed "Select all").
image

Filter (if desired). In this case I put in <= 1000 GRC. You don't HAVE to filter separately, because the consolidate will RE filter, but you can use the filter without consolidate if desired for other purposes.
image

Here is the consolidate destination address pick list. Note it does NOT show a 200 input limit warning, because the filter already limited the number of inputs to 200.
image

If you had gone directly from the select all to the consolidate, or modified the selections after the filter to add more inputs manually than 200. you would get this...
image

After you press ok to acknowledge coin control, the send tab shows the ready to send transaction with everything already filled in. All you have to do is press the send button.
image

**Comment by jamescowens** [comment]
So I think this is really good for the coin control version. Next up is a wizard, which I intend to reuse all of this machinery under the covers...

**Comment by jamescowens** [comment]
@RoboticMind ... something like
image

This now shows a green "ready to send" indicator with a tooltip.

**Comment by RoboticMind** [comment]
Might make the text a little brighter on the dark theme and maybe use the "ready to consolidate" wording instead, but I think something like that would work. Overall I think it's looking good

**Comment by jamescowens** [comment]
I have used a different green on the light vs dark stylesheet...
image
image

This should be ready to go now.

**Review by RoboticMind — APPROVED** [review]
Think it looks good now

File: src/qt/forms/coincontroldialog.ui [hunk 1]
```diff
@@ -285,90 +285,149 @@
...
+     <item>
+      <widget class="QPushButton" name="selectAllPushButton">
+       <property name="sizePolicy">
+        <sizepolicy hsizetype="Maximum" vsizetype="Fixed">
+         <horstretch>0</horstretch>
+         <verstretch>0</verstretch>
+        </sizepolicy>
+       </property>
+       <property name="toolTip">
+        <string>Flip-flops between selecting all and selecting none.</string>
```
Comment (cyrossignol) [inline src/qt/forms/coincontroldialog.ui]:
```suggestion
        <string>Toggles between selecting all and selecting none.</string>
```

👡👡

**Review by cyrossignol — COMMENTED** [review]
Could you please remind me about why we chose 200 outputs for the consolidation limit? Can't remember the discussion... been too long...

File: src/qt/coincontroldialog.cpp [hunk 2]
```diff
@@ -153,26 +165,194 @@ void CoinControlDialog::buttonBoxClicked(QAbstractButton* button)
...
+        ++iter;
     }
-    ui->treeWidget->setEnabled(false);
-    for (int i = 0; i < ui->treeWidget->topLevelItemCount(); i++)
-            if (ui->treeWidget->topLevelItem(i)->checkState(COLUMN_CHECKBOX) != state)
-                ui->treeWidget->topLevelItem(i)->setCheckState(COLUMN_CHECKBOX, state);
-    ui->treeWidget->setEnabled(true);
-    CoinControlDialog::updateLabels(model, this);
+
+    // The second loop is to limit the number of selected outputs to the inputCountLimit.
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coincontroldialog.cpp:330]: The original tests showed that consolidate transactions were failing above 200. It was also taking an extraordinary amount of time locking up the GUI because of the wallet core to GUI update triggers. The performance issue no longer exists because of the optimizaitons I did on the GUI update earlier, and this is originating in the GUI anyway. We could probably safely raise the output limits for both the rpc and here to 600.

[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coincontroldialog.cpp:330]: We could also do it based on size, although that is an estimate and is not completely reliable.

File: src/qt/coincontroldialog.cpp [hunk 3]
```diff
@@ -153,26 +165,194 @@ void CoinControlDialog::buttonBoxClicked(QAbstractButton* button)
...
+        m_ToState = Qt::Unchecked;
+    }
+    else
+    {
+        m_ToState = Qt::Checked;
+    }
+
+    if (m_ToState == Qt::Checked)
+    {
+       ui->selectAllPushButton->setText("Select all");
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coincontroldialog.cpp]: That is not consistent with the capitalization of the original screen, but I agree we should use all caps on buttons.

File: src/qt/coincontroldialog.cpp [hunk 4]
```diff
@@ -153,26 +165,194 @@ void CoinControlDialog::buttonBoxClicked(QAbstractButton* button)
...
+    }
+    else
+    {
+       ui->selectAllPushButton->setText("Select none");
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coincontroldialog.cpp]: ditto.

File: src/qt/coincontroldialog.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coincontroldialog.cpp:330]: I would rather not allow the filter to permit more than the number of outputs that could be successfully sent. I think the right answer is to raise the output limit fixed value.

File: src/qt/coincontroldialog.cpp [hunk 5]
```diff
@@ -153,26 +165,194 @@ void CoinControlDialog::buttonBoxClicked(QAbstractButton* button)
...
+
+    bool maxMinOutputValueValid = false;
+
+    outputFilterValue = ui->maxMinOutputValue->value(&maxMinOutputValueValid);
+
+    // If the Bitcoin amount field is not filled in for the filter, then fix up the value to max money for filter mode less,
+    // and zero for filter mode greater than.
+    if (!maxMinOutputValueValid) outputFilterValue = m_FilterMode ? MAX_MONEY: 0;
+
+    culled_inputs = filterInputsByValue(m_FilterMode, outputFilterValue, 200);
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coincontroldialog.cpp]: The problem with that approach is that someone may put and amount in and skip the filter and push the consolidate button directly, and that might also lead to unexpected results. We could solve that problem by greying out (disabling) the consolidate button until the filter button is pressed first, but that then requires people to use both the filter and consolidate button.

File: src/qt/forms/coincontroldialog.ui (hunk 1 above)
Comment (jamescowens) [reply src/qt/forms/coincontroldialog.ui]: Agreed.

File: src/qt/coincontroldialog.cpp [hunk 6]
```diff
@@ -153,26 +165,194 @@ void CoinControlDialog::buttonBoxClicked(QAbstractButton* button)
...
+    }
+
+    // If the number of inputs selected was limited, then true is returned.
+    return culled_inputs;
+}
+
+void CoinControlDialog::buttonConsolidateClicked()
+{
+    ConsolidateUnspentDialog* consolidateUnspentDialog = new ConsolidateUnspentDialog(this);
+    consolidateUnspentDialog->SetOutputWarningVisible(false);
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coincontroldialog.cpp]: Agreed.

File: src/qt/coincontroldialog.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coincontroldialog.cpp:330]: Ugh. After looking at this, I agree with you. I have removed the output count limitation from the filter button.

File: src/qt/coincontroldialog.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coincontroldialog.cpp]: Conjugately to the above, I have removed the value part of the filter and only enforce the input count limit for the consolidate button. In an attempt to preserve the ability to shortcut to the consolidate button without pressing the filter, I have retained the ? MAX_MONEY : 0: ternary, and only disable the consolidate button IF a valid amount has been placed in the filter field AND the filter button has NOT been pressed.

**Review by cyrossignol — APPROVED** [review]
Looks good.

**Comment by vincehusky** [comment]
👏
